File: mcdowell/php/ch2-linked-list.php
<?php

class LinkedList {
    public $first;

    public function addValue($val) {
        $node = new LinkedListNode($val);
        if (!$this->first) {
            $this->first = $node;
        }
        else {
            $this->first->addLinkedListNode($node);
        }
        return $node;
    }

    public function removeValue($val) {
        if (!$this->first) return 0;

        if ($this->first->val == $val) {
            $this->first = null;
            return 1;
        }

        return $this->first->removeValue($val, $this->first);
    }

    public function getAll() {
        $values = array();
        $pointer = $this->first;
        while ($pointer != null) {
            $values[] = $pointer->val;
            $pointer = $pointer->next;
        }
        return $values;
    }
}


class LinkedListNode {
    public $next;
    public $val;

    public function __construct($val) {
        $this->val = $val;
    }

    public function addLinkedListNode(LinkedListNode $node) {
        $pointer = $this;
        while ($pointer->next != null) {
            $pointer = $pointer->next;
        }
        $pointer->next = $node;
    }

    public function removeValue($val) {

        $pointer = $this;
        $lastPointer = null;
        $flag = 0;
        while ($pointer != null) {
            if ($pointer->val == $val) {
                $lastPointer->next = $pointer->next;
                $flag ++;
            }
            else {
                $lastPointer = $pointer;
            }
            $pointer = $pointer->next;
        }


        return $flag;
    }

    public function __toString() {
        return (string) $this->val;
    }
}


//$list = new LinkedList();
//$list->addValue(2);
//$list->addValue(3);
//$list->addValue(3);
//$list->addValue(3);
//$list->addValue(4);
//$list->addValue(5);
//$list->addValue(6);
//print_r($list->getAll());
//
//var_dump($list->removeValue(3));
//var_dump($list->removeValue(3));
//var_dump($list->removeValue(5));
//
//print_r($list->getAll());

File: mcdowell/python/ch2_linked_list.py

class LinkedList:
    first = None

    def addValue(self, val):
        node = LinkedListNode(val)
        if self.first is None:
            self.first = node
        else:
            self.first.addNode(node)
        return node

    def removeValue(self, val):
        if self.first is None:
            return 0
        elif self.first.val == val:
            self.first = None
            return 1
        else:
            return self.first.removeValue(val)

    def getAll(self):
        values = list()
        pointer = self.first
        while pointer is not None:
            values.append(pointer.val)
            pointer = pointer.next
        return values


class LinkedListNode:
    val = None
    next = None
    def __init__(self, val):
        self.val = val

    def addNode(self, node):
        pointer = self
        while pointer.next is not None:
            pointer = pointer.next
        pointer.next = node


    def removeValue(self, val):
        pointer = self
        lastPointer = None
        flag = 0

        while pointer is not None :
            if pointer.val == val:
                lastPointer.next = pointer.next
                flag += 1
            else:
                lastPointer = pointer
            pointer = pointer.next

        return flag

    def __str__(self):
        return str(self.val)


def test():
    l = LinkedList()
    l.addValue(1)
    l.addValue(2)
    l.addValue(3)
    l.addValue(3)
    l.addValue(3)
    l.addValue(4)
    l.addValue(5)

    print l.getAll()

    print l.removeValue(2)
    print l.removeValue(3)
    print l.removeValue(9)

    print l.getAll()



if __name__ == '__main__':
    test()